src/backend/distributed/commands/vacuum.c

```diff
@@ -73,13 +79,68 @@ static CitusVacuumParams VacuumStmtParams(VacuumStmt *vacstmt);
 void
 PostprocessVacuumStmt(VacuumStmt *vacuumStmt, const char *vacuumCommand)
 {
-	int relationIndex = 0;
-	List *vacuumRelationList = ExtractVacuumTargetRels(vacuumStmt);
-	List *relationIdList = NIL;
 	CitusVacuumParams vacuumParams = VacuumStmtParams(vacuumStmt);
+
+	List *relationIdList = VacuumRelationIdList(vacuumStmt, vacuumParams);
+
+	if (ContainsOnlyPostgresTables(relationIdList))
+	{
+		return;
+	}
+
+	if (vacuumParams.options & VACOPT_VACUUM)
+	{
+		/*
+		 * We commit the current transaction here so that the global lock
+		 * taken from the shell table for VACUUM is released, which would block execution
+		 * of shard placements.
+		 */
+		CommitTransactionCommand();
+		StartTransactionCommand();
+	}
+
+	/*
+	 * Here we get the relation list again because we might have
+	 * closed the current transaction.
+	 */
+	relationIdList = VacuumRelationIdList(vacuumStmt, vacuumParams);
+	bool distributedVacuumStmt = IsDistributedVacuumStmt(vacuumParams.options,
+														 relationIdList);
+	if (!distributedVacuumStmt)
+	{
+		return;
+	}
+
+	ExecuteVacuumOnDistributedTables(vacuumStmt, relationIdList, vacuumParams);
+}
+
+
+/*
+ * VacuumRelationIdList returns the oid of the relations in the given vacuum statement.
+ */
+static List *
+VacuumRelationIdList(VacuumStmt *vacuumStmt, CitusVacuumParams vacuumParams)
+{
 	LOCKMODE lockMode = (vacuumParams.options & VACOPT_FULL) ? AccessExclusiveLock :
 						ShareUpdateExclusiveLock;
-	int executedVacuumCount = 0;
+	const char *stmtName = (vacuumParams.options & VACOPT_VACUUM) ? "VACUUM" : "ANALYZE";
+
+	List *vacuumRelationList = ExtractVacuumTargetRels(vacuumStmt);
+
+	/*
+	 * No table in the vacuum statement means vacuuming all relations
+	 * which is not supported by citus.
+	 */
+	int vacuumedRelationCount = list_length(vacuumRelationList);
+	if (vacuumedRelationCount == 0)
+	{
+		/* WARN for unqualified VACUUM commands */
+		ereport(WARNING, (errmsg("not propagating %s command to worker nodes", stmtName),
+						  errhint("Provide a specific table in order to %s "
+								  "distributed tables.", stmtName)));
+	}
```

Comment:
maybe move to the beginning of PostprocessVacuumStmt , this does not seem to belong in a function named VacuumRelationIdList

(although we should really just send vacuum to workers as well)